### DISCUSSION TITLE
Unused _Variables reported as such

### DISCUSSION THREAD

**Discussion by paulo-ferraz-oliveira** [AdRoll/rebar3_hank#66] [General]
Let's say I have this code

```erlang
test(_Config) -> ok.
```

`rebar3_hank` signals the parameter as unused. But I feel that naming unused variables (instead of simply using _) helps readability (I use it extensively).

What is the rationale behind the warning? (or is there an option to turn it off I didn't notice?)

**Comment by elbrujohalcon** [comment]
The idea of the rule is to spot parameters that are unused in all clauses of the function and therefore unnecessary. The function would work as well with one less argument.

**Reply by paulo-ferraz-oliveira** [reply]
Not if it's a callback function 😄. And you will get those (a LOT) for test files, I imagine...

**Comment by elbrujohalcon** [comment]
Look no further than [`http_uri:parse_host_port/4`](https://github.com/erlang/otp/blob/6378a0c825db64df91e01ee39e3a268f4ba050b7/lib/inets/src/http_lib/http_uri.erl#L257-L266).

That function has three clauses…
```erlang
parse_host_port(_Scheme, DefaultPort, <<"[", HostPort/binary>>, Opts) -> %ipv6
    …
parse_host_port(_Scheme, DefaultPort, "[" ++ HostPort, Opts) -> %ipv6
    …
parse_host_port(_Scheme, DefaultPort, HostPort, _Opts) ->
    …
```

The first argument (`_Scheme`) is ignored in all of them.

Just to be sure… this function is not a behavior callback.

It's actually only used [here](https://github.com/erlang/otp/blob/6378a0c825db64df91e01ee39e3a268f4ba050b7/lib/inets/src/http_lib/http_uri.erl#L229) and [here](https://github.com/erlang/otp/blob/6378a0c825db64df91e01ee39e3a268f4ba050b7/lib/inets/src/http_lib/http_uri.erl#L245).

So, since the first argument is useless, it can be removed. That would turn `parse_host_port` into…

```erlang
parse_host_port(DefaultPort, <<"[", HostPort/binary>>, Opts) -> %ipv6
    …
parse_host_port(DefaultPort, "[" ++ HostPort, Opts) -> %ipv6
    …
parse_host_port(DefaultPort, HostPort, _Opts) ->
    …
```

…and it will simplify `parse_uri_rest/4` just a tiny bit, by removing `Scheme` from the calls to `parse_uri_rest/3`.

Cool, isn't it? 😎

**Reply by paulo-ferraz-oliveira** [reply]
It is cool in that case, sure. But there are "behaviours" that aren't defined as such (e.g. a CT suite) - imagining you already take callbacks into account (might not be the case).

**Reply by elbrujohalcon** [reply]
We take callbacks into account, yes… and for common test suites, at least until it's defined as a proper behavior… you (the user) will have to ignore `unused_ignored_function_parameters` like this…

```erlang
-hank [unused_ignored_function_parameters].
```

Even more, we're (in the current version - `0.2.1`) ignoring way too much. We plan to adjust that in #40.

**Reply by paulo-ferraz-oliveira** [reply]
That's the option I wanted. As for the behaviour, [it's been merged](https://github.com/erlang/otp/pull/2794) but it'll take a while before it's ubiquitous.

**Comment by pbrudnick** [comment]
This rule name is confusing, it's the proposal for the new name (by @elbrujohalcon )
https://github.com/AdRoll/rebar3_hank/issues/69
